pkg/broker/receiver.go

```diff
@@ -98,14 +101,55 @@ func (r *Receiver) getTrigger(ctx context.Context, ref provisioners.ChannelRefer
 }
 
 func (r *Receiver) shouldSendMessage(t *eventingv1alpha1.TriggerSpec, m *provisioners.Message) bool {
-	// TODO More filtering!
-	if t.Type != Any && t.Type != m.Headers["Ce-Eventtype"] {
-		r.logger.Debug("Wrong type", zap.String("trigger.spec.type", t.Type), zap.String("message.type", m.Headers["Ce-Eventtype"]), zap.Any("m", m))
+	// This conversion to selector should be done only once, possibly upon creation of the trigger
+	// in case the filters are immutable. All validations should be performed then.
+	selector, err := r.buildSelector(t)
+	if err != nil {
+		r.logger.Error("Invalid selector for trigger spec", zap.Any("triggerSpec", t))
 		return false
 	}
-	if t.Source != "" && t.Source != m.Headers["Ce-Source"] {
-		r.logger.Debug("Wrong source", zap.String("trigger.spec.source", t.Source), zap.String("message.source", m.Headers["Ce-Source"]))
-		return false
+	matched := selector.Matches(labels.Set(m.Headers))
+	if !matched {
+		r.logger.Debug("Selector did not match message headers", zap.String("selector", selector.String()), zap.Any("headers", m.Headers))
+	}
+	return matched
+}
+
+func (r *Receiver) buildSelector(ts *eventingv1alpha1.TriggerSpec) (labels.Selector, error) {
```

Comment:
This seems like it will be hard to test and somewhat brittle. I'd skip this for MVP.

Reply:
Will do